File: ext/websocket_driver/message.c
#include "message.h"

wsd_Message *wsd_Message_create(wsd_Frame *frame)
{
    wsd_Message *message = calloc(1, sizeof(wsd_Message));
    if (message == NULL) return NULL;

    message->frames = wsd_Queue_create();
    if (message->frames == NULL) {
        free(message);
        return NULL;
    }

    message->length = 0;
    if (!wsd_Message_push_frame(message, frame)) {
        wsd_clear_pointer(wsd_Queue_destroy, message->frames);
        free(message);
        return NULL;
    }

    message->opcode = frame->opcode;
    message->rsv1   = frame->rsv1;
    message->rsv2   = frame->rsv2;
    message->rsv3   = frame->rsv3;

    return message;
}

void wsd_Message_destroy(wsd_Message *message)
{
    if (message == NULL) return;

    wsd_Queue_each(message->frames, (wsd_Queue_cb)wsd_Frame_destroy);

    wsd_clear_pointer(wsd_Queue_destroy, message->frames);

    free(message);
}

int wsd_Message_push_frame(wsd_Message *message, wsd_Frame *frame)
{
    if (!wsd_Queue_push(message->frames, frame)) return 0;

    message->length += frame->length;
    return 1;
}

uint64_t wsd_Message_copy(wsd_Message *message, uint8_t *target)
{
    uint64_t offset = 0;
    wsd_QueueIter *iter = wsd_QueueIter_create(message->frames);

    if (iter == NULL) return 0;

    for (; iter->value; wsd_QueueIter_next(iter)) {
        offset = wsd_Frame_copy(iter->value, target, offset);
    }

    wsd_QueueIter_destroy(iter);

    return offset;
}

File: ext/websocket_driver/com/jcoglan/websocket/Message.java
package com.jcoglan.websocket;

import java.util.ArrayList;
import java.util.List;

public class Message {
    static int MAX_LENGTH = 0x3ffffff;

    public int opcode;
    public boolean rsv1;
    public boolean rsv2;
    public boolean rsv3;
    int length;
    List<Frame> frames;

    static boolean wouldOverflow(Message message, Frame frame) {
        int mlen  = (message == null) ? 0 : message.length;
        long flen = frame.length;

        return flen < 0 || flen > MAX_LENGTH - mlen;
    }

    Message() {
        length = 0;
        frames = new ArrayList<Frame>();
    }

    Message(Frame frame) {
        this();
        push(frame);

        opcode = frame.opcode;
        rsv1   = frame.rsv1;
        rsv2   = frame.rsv2;
        rsv3   = frame.rsv3;
    }

    void push(Frame frame) {
        length += frame.length;
        frames.add(frame);
    }

    public byte[] copy() {
        byte[] target = new byte[length];
        int offset = 0;

        for (Frame frame : frames) {
            System.arraycopy(frame.payload, 0, target, offset, (int)frame.length);
            offset += frame.length;
        }

        return target;
    }
}
